src/main/java/org/candlepin/model/CertificateSerialCurator.java

```diff
@@ -216,4 +85,46 @@ public int deleteRevokedExpiredAndNotCollectedSerials() {
             .setParameter("nowDate", new Date())
             .getResultList();
     }
+
+    /**
+     * Returns all serial ids that are revoked but not expired.
+     *
+     * @return a list of serial ids
+     */
+    @SuppressWarnings("unchecked")
+    public List<Long> listNonExpiredRevokedSerialIds() {
+        String hql = "SELECT s.id" +
+            "    FROM CertificateSerial s" +
+            "    WHERE" +
+            "       s.revoked=true" +
+            "    AND" +
+            "       s.expiration >= :nowDate";
+
+        Query query = this.getEntityManager().createQuery(hql);
+
+        return (List<Long>) query
+            .setParameter("nowDate", new Date())
+            .getResultList();
+    }
+
+    /**
+     * Revokes all serials specified by the given ids
+     *
+     * @param serialsToRevoke ids of the serials to be revoked
+     * @return a number of revoked serials
+     */
+    @Transactional
+    public int revokeByIds(Collection<Long> serialsToRevoke) {
+        if (serialsToRevoke == null || serialsToRevoke.isEmpty()) {
+            return 0;
+        }
+
+        String hql = "UPDATE CertificateSerial s SET s.revoked = true" +
+            " WHERE s.revoked = false AND s.id IN (:serials)";
```

Comment:
Need protection for the IN operator here.

Reply:
Done.